### PR TITLE
Add client-side credential reload to spiffe security connector

### PR DESCRIPTION
**Pull request by yihuazhang**
This PR adds the client-side credential reload support to SPIFFE stack that will fix https://github.com/grpc/grpc/issues/15735, https://github.com/grpc/grpc/issues/19237

It additionally made the following changes: 

1. Make each security connector to have it's own copy of  `grpc_tls_key_materials_config` instance which will be initially copied from `grpc_tls_credentials_options` instance. 
2. Add a version number to `grpc_tls_key_materials_config` struct to indicate the version of key materials. It will simplify the application's implementation to determinine if the key materials need to be reloaded.  
3. Use a mutex to prevent concurrent calls to `Schedule` API.

### CODE REVIEW

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h [hunk 1]
```diff
@@ -39,12 +39,15 @@ struct grpc_tls_key_materials_config
   const PemKeyCertPairList& pem_key_cert_pair_list() const {
     return pem_key_cert_pair_list_;
   }
+  int version() const { return version_; }
```
Comment (euroelessar) [inline src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h]: I assume this API should be exposed in `include/grpc/grpc_security.h`?

Comment (yihuazhang) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h]: Yes, that's correct. It should be exposed in `grpc_security.h`.

**Comment by euroelessar** [comment]
@jiangtaoli2016 @yihuazhang hi, any updates here?

**Comment by yihuazhang** [comment]
Sorry @euroelessar, I dropped the ball. I will add some tests and mail it out today.

File: src/core/lib/security/security_connector/tls/spiffe_security_connector.cc [hunk 2]
```diff
@@ -384,39 +454,71 @@ SpiffeServerSecurityConnector::CreateSpiffeServerSecurityConnector(
...
+  return ReplaceHandshakerFactory();
+}
+
+grpc_security_status SpiffeServerSecurityConnector::RefreshHandshakerFactory() {
+  grpc_core::MutexLock lock(&mu_);
+  const SpiffeServerCredentials* creds =
+      static_cast<const SpiffeServerCredentials*>(server_creds());
+  grpc_ssl_certificate_config_reload_status reload_status =
+      GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED;
+  FetchKeyMaterials(key_materials_config_, creds->options(), &reload_status);
```
Comment (jiangtaoli2016) [inline src/core/lib/security/security_connector/tls/spiffe_security_connector.cc]: the return status of FetchKeyMaterials does not populate.

File: src/core/lib/security/security_connector/tls/spiffe_security_connector.cc [hunk 3]
```diff
@@ -239,32 +260,72 @@ SpiffeChannelSecurityConnector::CreateSpiffeChannelSecurityConnector(
...
+}
+
+grpc_security_status
+SpiffeChannelSecurityConnector::RefreshHandshakerFactory() {
+  grpc_core::MutexLock lock(&mu_);
+  const SpiffeCredentials* creds =
+      static_cast<const SpiffeCredentials*>(channel_creds());
+  grpc_ssl_certificate_config_reload_status reload_status =
+      GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED;
+  FetchKeyMaterials(key_materials_config_, creds->options(), &reload_status);
```
Comment (jiangtaoli2016) [inline src/core/lib/security/security_connector/tls/spiffe_security_connector.cc]: the return status of FetchKeyMaterials does not populate.

File: include/grpc/grpc_security.h [hunk 4]
```diff
@@ -793,10 +808,11 @@ typedef void (*grpc_tls_on_credential_reload_done_cb)(
 /** A struct containing all information necessary to schedule/cancel
     a credential reload request. cb and cb_user_data represent a gRPC-provided
     callback and an argument passed to it. key_materials is an in/output
-    parameter containing currently used/newly reloaded credentials. status and
-    error_details are used to hold information about errors occurred when a
-    credential reload request is scheduled/cancelled. It is used for
-    experimental purpose for now and subject to change. */
+    parameter containing currently used/newly reloaded credentials. If
+   credential reload does not result in a new credential, key_materials should
```
[the remark this answers is not in the document]
Comment (yihuazhang) [reply include/grpc/grpc_security.h]: Done.

File: src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h [hunk 5]
```diff
@@ -39,12 +39,15 @@ struct grpc_tls_key_materials_config
...
 
   /** Setters for member fields. **/
   void set_key_materials(grpc_core::UniquePtr<char> pem_root_certs,
                          PemKeyCertPairList pem_key_cert_pair_list);
+  void set_version(int version) { version_ = version; }
 
  private:
+  int version_;
```
[the remark this answers is not in the document]
Comment (yihuazhang) [reply src/core/lib/security/credentials/tls/grpc_tls_credentials_options.h]: Done.

File: src/core/lib/security/security_connector/ssl_utils.h [hunk 6]
```diff
@@ -149,9 +149,19 @@ class PemKeyCertPair {
...
 
-  // Not copyable.
-  PemKeyCertPair(const PemKeyCertPair&) = delete;
-  PemKeyCertPair& operator=(const PemKeyCertPair&) = delete;
+  // Copyable.
+  PemKeyCertPair(const PemKeyCertPair& other)
+      : private_key_(gpr_strdup(other.private_key())),
+        cert_chain_(gpr_strdup(other.cert_chain())) {}
+  PemKeyCertPair& operator=(const PemKeyCertPair& other) {
+    auto private_key =
```
[the remark this answers is not in the document]
Comment (yihuazhang) [reply src/core/lib/security/security_connector/ssl_utils.h]: Good point. Done.

File: src/core/lib/security/security_connector/tls/spiffe_security_connector.cc (hunk 2 above)
Comment (yihuazhang) [reply src/core/lib/security/security_connector/tls/spiffe_security_connector.cc]: The return status is checked now.

File: src/core/lib/security/security_connector/tls/spiffe_security_connector.cc (hunk 3 above)
Comment (yihuazhang) [reply src/core/lib/security/security_connector/tls/spiffe_security_connector.cc]: The return status is checked now.

File: src/core/lib/security/security_connector/tls/spiffe_security_connector.h [hunk 7]
```diff
@@ -113,11 +125,28 @@ class SpiffeServerSecurityConnector final
...
-  grpc_security_status RefreshServerHandshakerFactory();
+  grpc_security_status RefreshHandshakerFactory();
+
+  grpc_core::Mutex mu_;
   tsi_ssl_server_handshaker_factory* server_handshaker_factory_ = nullptr;
+  grpc_core::RefCountedPtr<grpc_tls_key_materials_config> key_materials_config_;
 };
 
+// Exposed for testing only.
+grpc_status_code FetchKeyMaterials(
```
[the remark this answers is not in the document]
Comment (yihuazhang) [reply src/core/lib/security/security_connector/tls/spiffe_security_connector.h]: Done. 

I place `TlsFetchKeyMaterials` inside `grpc_core::internal` namespace. Not sure if the whole file needs to be in `grpc_core` namespace. I checked ssl, alts, and local, and non of them is placed inside grpc_core namespace. HDYT?

**Comment by yihuazhang** [comment]
https://github.com/grpc/grpc/issues/19669,https://github.com/grpc/grpc/issues/19802,https://github.com/grpc/grpc/issues/19819

**Comment by yihuazhang** [comment]
https://github.com/grpc/grpc/issues/19819
https://github.com/grpc/grpc/issues/19802
https://github.com/grpc/grpc/issues/19669

**Pull request merged**